src/docfx/restore/RestoreLocker.cs

```diff
@@ -0,0 +1,89 @@
+        private readonly RestoreLock _restoreItem;
+
+        public RestoreLocker(string docset)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(docset));
+
+            _restoreItem = Load(docset).Result;
+        }
+
+        public bool TryGetWorkTreeHead(string href, out string workTreeHead)
+        {
+            return _restoreItem.Git.TryGetValue(href, out workTreeHead);
+        }
+
+        public static Task Lock(string docset, Func<RestoreLock, RestoreLock> process)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(docset));
+
+            var restoreLockFilePath = GetRestoreLockFilePath(docset);
+            return ProcessUtility.ProcessLock(
+                Path.GetRelativePath(AppData.RestoreLockDir, restoreLockFilePath),
+                () =>
+                {
+                    Directory.CreateDirectory(Path.GetDirectoryName(restoreLockFilePath));
+                    using (var fileStream = File.Open(restoreLockFilePath, FileMode.OpenOrCreate, FileAccess.ReadWrite))
+                    {
+                        var sr = new StreamReader(fileStream);
+                        var sw = new StreamWriter(fileStream);
+
+                        // read restore item
+                        var (_, restoreItem) = JsonUtility.Deserialize<RestoreLock>(sr.ReadToEnd());
+                        if (restoreItem == null)
+                            restoreItem = new RestoreLock();
+
+                        // process restore item
+                        restoreItem = process(restoreItem);
+
+                        // write back restore item
+                        fileStream.SetLength(0);
+                        sw.Write(JsonUtility.Serialize(restoreItem));
+                        sw.Flush();
+
+                        return Task.CompletedTask;
+                    }
+                });
+        }
+
+        private static async Task<RestoreLock> Load(string docset)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(docset));
+
+            var restoreLockFilePath = GetRestoreLockFilePath(docset);
+            var restore = new RestoreLock();
+            await ProcessUtility.ProcessLock(
+                Path.GetRelativePath(AppData.RestoreLockDir, restoreLockFilePath),
+                () =>
+                {
+                    if (File.Exists(restoreLockFilePath))
+                    {
+                        (_, restore) = JsonUtility.Deserialize<RestoreLock>(File.ReadAllText(restoreLockFilePath));
+                    }
+
+                    return Task.CompletedTask;
+                });
+
+            return restore;
+        }
+
+        private static string GetRestoreLockFilePath(string docset)
+        {
+            docset = PathUtility.NormalizeFile(Path.GetFullPath(docset));
+            var docsetKey = Path.GetFileName(docset) + "-" + PathUtility.Encode(HashUtility.GetMd5String(docset));
+
+            return Path.Combine(AppData.RestoreLockDir, $"{docsetKey}-lock.json");
```

Comment:
>Path.Combine [](start = 19, length = 12)

Move Path.Combine into ProcessLock's action? to avoid an extra Path.GetRelativePath to get restore lock file name.

Reply:
ProcessLock and RestoreLock are different...

---
In reply to: [200908463](https://github.com/dotnet/docfx/pull/3020#discussion_r200908463) [](ancestors = 200908463)

Reply:
Can we do some rename (maybe later) to avoid the confusion? how about `RestoreLockFile` and `FileMutex`?

- A lock file is more likely to relate to similar concepts: `package-lock.json`, `project.lock.json`, `yarn.lock`, `pod.lock`...
- Mutex is a well known synchronization primitve, and `FileMutex` suggests it is based on file system, thus can handle inter-process locking

Reply:
Good suggestions, I will rename them in another PR

---
In reply to: [200910345](https://github.com/dotnet/docfx/pull/3020#discussion_r200910345) [](ancestors = 200910345)